File: day12/zig/src/main.zig
const std = @import("std");

const Cell = struct { row: i32, col: i32 };

const Bounds = struct { width: i32, height: i32 };

const Shape = struct {
    cells: std.AutoHashMap(Cell, void),

    fn deinit(self: *Shape) void {
        self.cells.deinit();
    }
};

const Region = struct {
    width: i32,
    height: i32,
    shape_counts: std.ArrayList(i32),

    fn deinit(self: *Region, allocator: std.mem.Allocator) void {
        self.shape_counts.deinit(allocator);
    }
};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var stdout_buf: [256]u8 = undefined;
    var stdout_writer = std.fs.File.stdout().writer(&stdout_buf);
    const stdout = &stdout_writer.interface;

    const file = try std.fs.cwd().openFile("../input.txt", .{});
    defer file.close();

    const content = try file.readToEndAlloc(allocator, 1024 * 1024);
    defer allocator.free(content);

    var lines: std.ArrayList([]const u8) = .empty;
    defer lines.deinit(allocator);

    var line_iter = std.mem.splitScalar(u8, content, '\n');
    while (line_iter.next()) |line| {
        try lines.append(allocator, line);
    }

    var shapes: std.ArrayList(Shape) = .empty;
    defer {
        for (shapes.items) |*s| s.deinit();
        shapes.deinit(allocator);
    }

    var regions: std.ArrayList(Region) = .empty;
    defer {
        for (regions.items) |*r| r.deinit(allocator);
        regions.deinit(allocator);
    }

    var index: usize = 0;
    while (index < lines.items.len) {
        const line = lines.items[index];

        // Check for shape definition (digit followed by colon, no 'x')
        if (line.len > 0 and std.ascii.isDigit(line[0]) and
            std.mem.indexOfScalar(u8, line, ':') != null and
            std.mem.indexOfScalar(u8, line, 'x') == null)
        {
            var shape_lines: std.ArrayList([]const u8) = .empty;
            defer shape_lines.deinit(allocator);

            index += 1;
            while (index < lines.items.len) {
                const shape_line = lines.items[index];
                if (shape_line.len == 0) break;
                if (std.mem.indexOfScalar(u8, shape_line, 'x') != null) break;
                if (shape_line.len > 0 and std.ascii.isDigit(shape_line[0]) and
                    std.mem.indexOfScalar(u8, shape_line, ':') != null)
                {
                    break;
                }
                try shape_lines.append(allocator, shape_line);
                index += 1;
            }

            var cells = std.AutoHashMap(Cell, void).init(allocator);
            for (shape_lines.items, 0..) |row, row_idx| {
                for (row, 0..) |ch, col_idx| {
                    if (ch == '#') {
                        try cells.put(.{ .row = @intCast(row_idx), .col = @intCast(col_idx) }, {});
                    }
                }
            }
            try shapes.append(allocator, Shape{ .cells = cells });
        } else if (std.mem.indexOfScalar(u8, line, 'x') != null and std.mem.indexOfScalar(u8, line, ':') != null) {
            // Region definition
            var parts = std.mem.splitSequence(u8, line, ": ");
            const dimensions_str = parts.next().?;
            const counts_str = parts.next().?;

            var dim_parts = std.mem.splitScalar(u8, dimensions_str, 'x');
            const width = try std.fmt.parseInt(i32, dim_parts.next().?, 10);
            const height = try std.fmt.parseInt(i32, dim_parts.next().?, 10);

            var shape_counts: std.ArrayList(i32) = .empty;
            var count_iter = std.mem.splitScalar(u8, counts_str, ' ');
            while (count_iter.next()) |count_str| {
                const trimmed = std.mem.trim(u8, count_str, &std.ascii.whitespace);
                if (trimmed.len > 0) {
                    try shape_counts.append(allocator, try std.fmt.parseInt(i32, trimmed, 10));
                }
            }

            try regions.append(allocator, Region{
                .width = width,
                .height = height,
                .shape_counts = shape_counts,
            });
            index += 1;
        } else {
            index += 1;
        }
    }

    // Calculate shape properties
    var shape_cell_counts = try allocator.alloc(i32, shapes.items.len);
    defer allocator.free(shape_cell_counts);

    var shape_bounds = try allocator.alloc(Bounds, shapes.items.len);
    defer allocator.free(shape_bounds);

    for (shapes.items, 0..) |shape, i| {
        shape_cell_counts[i] = @intCast(shape.cells.count());
        shape_bounds[i] = boundingBox(&shape);
    }

    var fit_count: u32 = 0;
    for (regions.items) |region| {
        if (canFitPresents(region, shape_cell_counts, shape_bounds)) {
            fit_count += 1;
        }
    }

    try stdout.print("Part 1: {d}\n", .{fit_count});
    try stdout.print("Part 2: Merry Christmas!\n", .{});
    try stdout.flush();
}

fn boundingBox(shape: *const Shape) Bounds {
    if (shape.cells.count() == 0) return Bounds{ .width = 0, .height = 0 };

    var min_row: i32 = std.math.maxInt(i32);
    var max_row: i32 = std.math.minInt(i32);
    var min_col: i32 = std.math.maxInt(i32);
    var max_col: i32 = std.math.minInt(i32);

    var iter = shape.cells.keyIterator();
    while (iter.next()) |cell| {
        if (cell.row < min_row) min_row = cell.row;
        if (cell.row > max_row) max_row = cell.row;
        if (cell.col < min_col) min_col = cell.col;
        if (cell.col > max_col) max_col = cell.col;
    }

    return .{
        .width = max_row - min_row + 1,
        .height = max_col - min_col + 1,
    };
}

fn canFitPresents(
    region: Region,
    shape_cell_counts: []const i32,
    shape_bounds: []const Bounds,
) bool {
    // Check total area
    var total_cells_needed: i64 = 0;
    for (region.shape_counts.items, 0..) |count, i| {
        if (i < shape_cell_counts.len) {
            total_cells_needed += @as(i64, count) * @as(i64, shape_cell_counts[i]);
        }
    }

    const region_area: i64 = @as(i64, region.width) * @as(i64, region.height);
    if (total_cells_needed > region_area) return false;

    // Check bounding boxes
    for (region.shape_counts.items, 0..) |count, i| {
        if (i >= shape_bounds.len) continue;
        if (count == 0) continue;

        const shape_w = shape_bounds[i].width;
        const shape_h = shape_bounds[i].height;

        const fits_normal = shape_w <= region.width and shape_h <= region.height;
        const fits_rotated = shape_h <= region.width and shape_w <= region.height;

        if (!fits_normal and !fits_rotated) return false;
    }

    return true;
}

File: day12/rust/src/main.rs
use std::collections::HashSet;
use std::fs;

struct Shape {
    cells: HashSet<(i32, i32)>,
}

struct Region {
    width: i32,
    height: i32,
    shape_counts: Vec<i32>,
}

fn main() {
    let input = fs::read_to_string("../input.txt").expect("Failed to read input file");
    let lines: Vec<&str> = input.lines().collect();
    let (shapes, regions) = parse_input(&lines);

    let shape_cell_counts: Vec<i32> = shapes.iter().map(count_cells).collect();
    let shape_bounds: Vec<(i32, i32)> = shapes.iter().map(bounding_box).collect();

    let fit_count = regions
        .iter()
        .filter(|region| can_fit_presents(region, &shape_cell_counts, &shape_bounds))
        .count();

    println!("Part 1: {}", fit_count);
    println!("Part 2: Merry Christmas!");
}

fn parse_input(lines: &[&str]) -> (Vec<Shape>, Vec<Region>) {
    let mut shapes = Vec::new();
    let mut regions = Vec::new();
    let mut index = 0;

    while index < lines.len() {
        let line = lines[index];

        if line.chars().next().map_or(false, |c| c.is_ascii_digit())
            && line.ends_with(':')
            && !line.contains('x')
        {
            let mut shape_lines = Vec::new();
            index += 1;

            while index < lines.len()
                && !lines[index].is_empty()
                && !lines[index].contains('x')
                && !(lines[index].chars().next().map_or(false, |c| c.is_ascii_digit())
                    && lines[index].ends_with(':'))
            {
                shape_lines.push(lines[index]);
                index += 1;
            }

            let mut cells = HashSet::new();
            for (row_index, row) in shape_lines.iter().enumerate() {
                for (col_index, ch) in row.chars().enumerate() {
                    if ch == '#' {
                        cells.insert((row_index as i32, col_index as i32));
                    }
                }
            }
            shapes.push(Shape { cells });
        } else if line.contains('x') && line.contains(':') {
            let parts: Vec<&str> = line.split(": ").collect();
            let dimensions: Vec<i32> = parts[0].split('x').map(|s| s.parse().unwrap()).collect();
            let width = dimensions[0];
            let height = dimensions[1];
            let shape_counts: Vec<i32> = parts[1].split(' ').map(|s| s.parse().unwrap()).collect();
            regions.push(Region {
                width,
                height,
                shape_counts,
            });
            index += 1;
        } else {
            index += 1;
        }
    }

    (shapes, regions)
}

fn count_cells(shape: &Shape) -> i32 {
    shape.cells.len() as i32
}

fn bounding_box(shape: &Shape) -> (i32, i32) {
    if shape.cells.is_empty() {
        return (0, 0);
    }

    let rows: Vec<i32> = shape.cells.iter().map(|(r, _)| *r).collect();
    let cols: Vec<i32> = shape.cells.iter().map(|(_, c)| *c).collect();

    let min_row = *rows.iter().min().unwrap();
    let max_row = *rows.iter().max().unwrap();
    let min_col = *cols.iter().min().unwrap();
    let max_col = *cols.iter().max().unwrap();

    (max_row - min_row + 1, max_col - min_col + 1)
}

fn can_fit_presents(
    region: &Region,
    shape_cell_counts: &[i32],
    shape_bounds: &[(i32, i32)],
) -> bool {
    // Check total area constraint
    let total_cells_needed: i64 = region
        .shape_counts
        .iter()
        .zip(shape_cell_counts.iter())
        .map(|(&count, &cells_per_shape)| count as i64 * cells_per_shape as i64)
        .sum();

    let region_area = region.width as i64 * region.height as i64;

    if total_cells_needed > region_area {
        return false;
    }

    // Check each shape's bounding box fits in region (allowing rotation)
    region
        .shape_counts
        .iter()
        .zip(shape_bounds.iter())
        .all(|(&count, &(shape_width, shape_height))| {
            count == 0
                || (shape_width <= region.width && shape_height <= region.height)
                || (shape_height <= region.width && shape_width <= region.height)
        })
}
